### PR TITLE
Add chaikins smoothing algorithm

### PR DESCRIPTION
**Pull request by nmandery**
This PR adds an implementation of [Chaikins smoothing](http://www.idav.ucdavis.edu/education/CAGDNotes/Chaikins-Algorithm/Chaikins-Algorithm.html) algorithm.

- [x] I agree to follow the project's [code of conduct](https://github.com/georust/geo/blob/master/CODE_OF_CONDUCT.md).
- [x] I added an entry to `CHANGES.md` if knowledge of this change could be valuable to users.

Thank you for creating these crates.

### CODE REVIEW

File: geo/src/algorithm/chaikins_smoothing.rs [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+{
+    let mut out_coords: Vec<_> = Vec::with_capacity(linestring.0.len() * 2);
+
+    if let (Some(first), Some(last)) = (linestring.0.first(), linestring.0.last()) {
+        if first != last {
+            // preserve start coordinate when the linestring is open
+            out_coords.push(*first);
+        }
+    }
+    for window_coordiates in linestring.0.windows(2) {
```
Comment (urschrei) [inline geo/src/algorithm/chaikins_smoothing.rs:97]: Typo fix: `window_coordiates` --> `window_coordinates`

File: geo/src/algorithm/chaikins_smoothing.rs [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+
+use crate::{CoordFloat, Coordinate, LineString, MultiLineString, MultiPolygon, Polygon};
+
+/// Smoothen `LineString`, `Polygon`, `MultiLineString` and `MultiPolygon` using Chaikins algorithm.
+///
+/// [Chaikins smoothing algorithm](http://www.idav.ucdavis.edu/education/CAGDNotes/Chaikins-Algorithm/Chaikins-Algorithm.html)
+///
+/// This implementation preserves the start and end vertices of an open linestring and
+/// smoothes the corner between start and end of a closed linestring.
+///
```
Comment (urschrei) [inline geo/src/algorithm/chaikins_smoothing.rs:13]: Would it make sense to add a line explaining the effect of increasing `n_iterations`?

**Review by urschrei — CHANGES_REQUESTED** [review]
Hi @nmandery, thanks for the PR! I've got some small changes, and one more general: while the algorithm is known as "Chaikin's smoothing algorithm" and it's fine to describe it like that in the docs, I'd prefer if the module and related functions were named e.g. `ChaikinSmoothing` and `chaikin_smoothing`, without the repeated "s", which is a bit unintuitive to remember / type.

File: geo/src/algorithm/chaikins_smoothing.rs (hunk 1 above)
Comment (nmandery) [reply geo/src/algorithm/chaikins_smoothing.rs]: Good catch - corrected

File: geo/src/algorithm/chaikins_smoothing.rs (hunk 2 above)
Comment (nmandery) [reply geo/src/algorithm/chaikins_smoothing.rs]: Also a good point - I added that and reworded that comment.

**Comment by nmandery** [comment]
Thanks for the feedback. I changed the naming - it is indeed nicer without the "s".

**Comment by rmanoka** [comment]
@nmandery Neat implementation!  Is it true that the number of vertices in the line string increases exponentially with number of iterations?  So m vertices becomes about `m*2^k` after k iterations?  This may be useful to document as it'll help the user pick reasonable k to not explode memory usage.  Also, you could consider initialising using  `Vec::with_capacity` if that's the case.

**Comment by nmandery** [comment]
@rmanoka Thank you. Yes each iteration doubles the number of vertices - I just added that to the documentation of the trait. Like you said, this is quite essential for the user to know.

I am currently using `Vec::with_capacity` to create the vec for the new `LineString` created during one iteration. The reserved size is twice the size of the input vec. It may be more efficient to rework the algorithm to reserve a new vec only once across all iterations and operate on that. Not sure if thats worth it, though.

**Comment by rmanoka** [comment]
> I am currently using `Vec::with_capacity` to create the vec for the new `LineString` created during one iteration. The reserved size is twice the size of the input vec. It may be more efficient to rework the algorithm to reserve a new vec only once across all iterations and operate on that. Not sure if thats worth it, though.

Yeah that's fair.  It's probably a micro optimization that should not be a bottle-neck at all.

**Review by rmanoka — APPROVED** [review]
Looks good to me

**Review by urschrei — APPROVED** [review]

**Pull request merged**